shared/src/TwinCollection.cs

```diff
@@ -289,10 +292,10 @@ private void TryClearMetadata(string propertyName)
         /// </summary>
         public void ClearMetadata()
         {
-            TryClearMetadata(MetadataName);
-            TryClearMetadata(LastUpdatedName);
-            TryClearMetadata(LastUpdatedVersionName);
-            TryClearMetadata(VersionName);
+            // GitHub Issue: https://github.com/Azure/azure-iot-sdk-csharp/issues/1971
+            // When we clear the metadata from the underlying collection we need to also clear
+            // the _metadata object so the TryGetMemberInternal will return a JObject instead of a new TwinCollection
+            _metadata.RemoveAll();
```

Comment:
What is the effective change in behavior between `TryClearMetadata(VersionName)` and this full clear? Might we ever expect a customer's app to be broken by this change?

Reply:
> What is the effective change in behavior between `TryClearMetadata(VersionName)` and this full clear? 

TryClearMetadata just removes the property from the underlying payload dictionary but doesn't touch the _metadata object. Unsure of why it does this but it's been this way for a while.

Currently the ClearMetadata() method only clears the top level metadata and does not clear the property level metadata. 

> Might we ever expect a customer's app to be broken by this change?

The behavior of this method **WILL** change. An example would be if someone who is used to the behavior clears the metadata and then attempts to cast the property to TwinCollection.

```c#
var twin = await registryManager.GetTwinAsync("device");
twin.Properties.Desired.ClearMetadata(); // New implementation
var actualProp = completeTwin.Properties.Reported[propName]; // Returns dynamic, but underlying type is JObject
actualProp.ToJson(); // Fails beacause the member doesn't exist.
```

Reply:
Perhaps instead we should add a new customer-facing method that does a full clear?

Reply:
Sounds reasonable.